sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuBroadcastToRowExec.scala

```diff
@@ -75,43 +80,51 @@ case class GpuBroadcastToRowExec(
 
   override def doCanonicalize(): SparkPlan = {
     val keys = buildKeys.map(k => QueryPlan.normalizeExpressions(k, child.output))
-    GpuBroadcastToRowExec(index, keys, child.canonicalized)(modeKeys)
+    GpuBroadcastToRowExec(keys, broadcastMode, child.canonicalized)(index, modeKeys)
   }
 
-  private def projectSerializedBatchToRows(
-      serBatch: SerializeConcatHostBuffersDeserializeBatch): Array[InternalRow] = {
+
+  private def projectSerializedBatch(
+      serBatch: SerializeConcatHostBuffersDeserializeBatch): Any = {
     val beforeCollect = System.nanoTime()
 
-    // Creates projection to extract target field from Row, as what Spark does.
+    // Creates projection to extract target fields from Row, as what Spark does.
     // Note that unlike Spark, the GPU broadcast data has not applied the key expressions from
     // the HashedRelation, so that is applied here if necessary to ensure the proper values
     // are being extracted. The CPU already has the key projections applied in the broadcast
     // data and thus does not have similar logic here.
     val broadcastModeProject = modeKeys.map { keyExprs =>
-      val keyExpr = keyExprs(index)
-      UnsafeProjection.create(keyExpr)
+      UnsafeProjection.create(keyExprs)
     }
 
-    // Use the single output of the broadcast mode projection if it exists
-    val rowProjectIndex = if (broadcastModeProject.isDefined) 0 else index
-    val rowProject = UnsafeProjection.create(
-      BoundReference(rowProjectIndex, buildKeys(index).dataType, buildKeys(index).nullable))
-
     // Deserializes the batch on the host. Then, transforms it to rows and performs row-wise
     // projection. We should NOT run any device operation on the driver node.
     val result = withResource(serBatch.hostBatches) { hostBatches =>
       hostBatches.flatMap { cb =>
         cb.rowIterator().asScala.map { row =>
           val broadcastRow = broadcastModeProject.map(_(row)).getOrElse(row)
-          rowProject(broadcastRow).copy().asInstanceOf[InternalRow]
+          broadcastMode match {
+            case map @ HashedRelationBroadcastMode(_, _) =>
+              broadcastRow.copy().asInstanceOf[InternalRow]
+            case _ =>
+              val idx = index.get
+              val rowProject = UnsafeProjection.create(
+                BoundReference(idx, buildKeys(idx).dataType, buildKeys(idx).nullable))
+              rowProject(broadcastRow).copy().asInstanceOf[InternalRow]
+          }
         }
       }
     }
 
     gpuLongMetric("dataSize") += serBatch.dataSize
     gpuLongMetric(COLLECT_TIME) += System.nanoTime() - beforeCollect
 
-    result
+    broadcastMode match {
+      case map @ HashedRelationBroadcastMode(_, _) =>
+        SparkShimImpl.broadcastModeTransform(map, result)
+      case _ =>
+        result
+    }
```

Comment:
Shouldn't this be simplified to just always call the broadcastModeTransform?  broadcastModeTransform already does nothing if it's an IdentityBroadcastMode.